### ISSUE TITLE
Use of "-frounding-math" (required by CGAL) causes constexpr to break.

### ISSUE THREAD

**Issue by richard-cms** [cms-sw/cmssw#14862]
I'm from the HI community. We keep most of our analysis code as a fork of CMSSW mainline, and I am attempting to port it from CMSSW_7_5_X to CMSSW_8_0_X. I've run into a bit of a pathological issue.

The RecoHI/HiJetAlgos/plugins/Buildfile.xml adds "-frounding-math" to the CXXFLAGS as required by the CGAL dependency (used by the Voronoi jet algorithm). Unfortunately starting in CMSSW_8_0_X this seems to break some constexpr involving Pi:

```shell
/cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_0_9/src/DataFormats/Math/interface/deltaPhi.h:16:34: error: '(1.0e+0 / 6.2831853071795862e+0)' is not a constant expression
    constexpr T o2pi = 1./(2.*M_PI);
                                  ^
```

This means that any code we add to the package in our analysis fork cannot compile if it includes basically any physics object headers, since they usually include deltaPhi.h in turn.

I have found this SO post explaining part of the issue:
http://stackoverflow.com/questions/20791333/constexpr-errors-with-floating-point-representation

Does anyone know of a workaround for this? 

@mandrenguyen, @mverwe, @dgulhan, @cfmcginn, and @yetkinyilmaz are interested in this as well.

**Comment by davidlange6** [comment]
@richard-cms - can you summarize my having a cgal dependency requires this flag?

**Comment by richard-cms** [comment]
@davidlange6 

The flag appears to be a hard dependency for CGAL - there is an error reported during compilation of that module if "-frounding-math" is removed. I don't know the details of why CGAL requires it. A quick google search shows that several projects are annoyed by this hard requirement.

**Comment by davidlt** [comment]
Googling suggests it's used by `CGAL::Interval_nt` and quick look at code suggest it has run-time checks. There is away to disable these run-time check. E.g. it's done for Intel C++ out-of-the-box. Here is the macro which controls it: `-DCGAL_DISABLE_ROUNDING_MATH_CHECK`. See `Number_types/include/CGAL/Interval_nt.h`. You also need to disable it if you want to use valgrind.

**Comment by richard-cms** [comment]
@davidlange6 Thanks for the find! We'll test it a bit to see if we can do without the rounding.

**Comment by richard-cms** [comment]
After testing for a while we think this is a working solution. Thanks again for the help, I'm closing the issue.
